source/module_base/formatter.h

```diff
@@ -1,43 +1,307 @@
-* providing from basic the single variable formatting to complex vectors, tables formatting output
-* Author: Yike HUANG, Daye ZHENG
-* Institution: AISI
-* ps. Zuxin JIN also helps solve dependency problems
-* ## Classes
-* - Fmt: single variable formatting (0 dimension)
-* - PhysicalFmt: bundle with Fmt object, work as a decorator, adjusting the Fmt object according to the physical context, energy, force, coordinate, etc.
-* - Table: output data structure, can be used to establish a table with titles, frames, etc.
-* - ContextFmt: context formatting, providing the most sophisticated formatting
-* ## Usage
-* There is a Chinese Feishu document available for this library, please refer to it for more details:
-* https://ucoyxk075n.feishu.cn/docx/Yym9dnm3aoTMfHxin8rcX9Rvnmb?theme=LIGHT&contentTheme=DARK
-* ## Structure
-* - File structure:
-* formatter.h: header file
-* formatter.cpp: source file
-* test/formatter_fmt_test.cpp: unit test for Fmt class
-* test/formatter_physfmt_test.cpp: unit test for PhysicalFmt class
-* test/formatter_table_test.cpp: unit test for Table class
-* test/formatter_contextfmt_test.cpp: unit test for ContextFmt class
-* - Class structure:
-*               Fmt: single variable formatting (0 dimension)
-*                |
-*                |
-*                |
-*          PhysicalFmt: decorator of Fmt
-*                |
-*                |
-*                |
-*           ContextFmt <----(inheritance)---- Table
-*/
-namespace formatter
+#include <string>
+#include "./ndarray.h"
+#include <iostream>
+#include <type_traits>
+#include <utility>
+/**
+ * @brief 
+ * 
+ * In C++20, the std::format library is introduced. However, it is not supported under restriction of ABACUS development that not later than C++11. Plus in ABACUS the formatting-output demands is not quite general but more specific, therefore, a simple alternative is proposed here.
+ * To use:
+ * 1. Use the static function format() to format data like `FmtCore::format("%d", 1);`
+ * 2. Use the class FmtCore to format data like `FmtCore fmt("%d"); fmt.format(1);`.
+ * The first way is more flexible while the second way is more efficient. The format string can be reset by reset() function. If empty, the format string is empty, otherwise it will be updated.
+ */
+class FmtCore
 {
-}
-/// @brief global variable, declared in module_base/formatter.h and defined in module_base/formatter_contextfmt.cpp for convenience, the same as INPUT
-extern formatter::ContextFmt context;
-#endif
+public:
+    FmtCore(const std::string& fmt): fmt_(fmt) {};
+    ~FmtCore() {};
+    /**
+     * @brief static function to format data
+     * 
+     * @tparam Ts datatype of the data
+     * @param fmt format string
+     * @param args data to format
+     * @return std::string 
+     */
+    template<typename... Ts>
+    static inline std::string format(const char* fmt, const Ts&... args)
+    {
+        int size = snprintf(nullptr, 0, fmt, FmtCore::filter(args)...) + 1;
```

Comment:
is this a const? const int size?

Reply:
thanks, I will add const later in this PR